msp/msp.go

```diff
@@ -0,0 +1,380 @@
+package msp
+
+import (
+	"container/heap"
+	"crypto/rand"
+	"errors"
+	"math/big"
+	"strings"
+)
+
+// A UserDatabase is an abstraction over the name -> share map returned by the
+// secret splitter that allows an application to only decrypt or request shares
+// when needed, rather than re-build a partial map of known data.
+type UserDatabase interface {
+	ValidUser(name string) bool
+	CanGetShare(string) bool
+	GetShare(string) ([][]byte, error)
+}
+
+type Condition interface { // Represents one condition in a predicate
+	Ok(*UserDatabase) bool
+}
+
+type String struct { // Type of condition
+	string
+	index int
```

Comment:
If you're reading a formatted or raw parse tree from left to right, the index is the number of times you've seen this user so far.  When we're recovering the secret, if we want to satisfy this node, we'll use the `index`th share of user `string`. (A user will have `n` shares if their name appears in the predicate `n` times.)